common/nym_offline_compact_ecash/src/scheme/mod.rs

```
@@ -0,0 +1,1114 @@
+    /// If the spending date is within the allowed range before the expiration date, the check is successful.
+    ///
+    /// This function performs two checks to ensure the payment expiration signature is valid:
+    /// - Verifies that the element `h` of the expiration signature does not equal the identity.
+    /// - Performs a bilinear pairing check involving the elements of the expiration signature and the payment (`h`, `kappa_e`, and `s`).
+    ///
+    /// # Arguments
+    ///
+    /// * `params` - A reference to the system parameters required for the checks.
+    /// * `verification_key` - The global verification key of the signing authorities.
+    /// * `spend_date` - The date associated with the payment.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` indicating success if the expiration signature is valid or an error if any check fails.
+    ///
+    /// # Errors
+    ///
+    /// An error is returned if:
+    /// - The element `h` of the payment expiration signature equals the identity.
+    /// - The bilinear pairing check for `kappa_e` fails.
+    ///
+    pub fn check_exp_signature_validity(
+        &self,
+        params: &Parameters,
+        verification_key: &VerificationKeyAuth,
+        spend_date: Scalar,
+    ) -> Result<()> {
+        // Check if the element h of the payment expiration signature equals the identity.
+        if bool::from(self.sig_exp.h.is_identity()) {
+            return Err(CompactEcashError::ExpirationDate(
+                "The element h of the payment expiration signature equals the identity".to_string(),
+            ));
+        }
+
+        // Calculate m1 and m2 values.
+        let m1: Scalar = spend_date;
+        let m2 = constants::TYPE_EXP;
+
+        // Perform a bilinear pairing check for kappa_e
+        let combined_kappa_e = self.kappa_e
+            + verification_key.beta_g2.get(1).unwrap() * m1
```

try to avoid using `unwrap` everywhere. if you're really certain this can't blow up (with the current API it absolutelye can since there's no bound checks on creation of `VerificationKeyAuth`), use `Index` api which hides the failure inside of it, i.e. `verification_key.beta_g2[2]`

---

Fixed in ebd882992abaa333d3f6a3d61411526ac4830a14